File: backend/routers/interview.py
from fastapi import APIRouter, Depends
from backend.models.user_model import UserProfile
from backend.models.interview_model import (
    InterviewStartRequest,
    InterviewMessageRequest,
    InterviewStartResponse,
    InterviewMessageResponse,
    InterviewSessionSummary,
    InterviewSessionDetail,
    InterviewMessage,
)
from backend.services.interview_service import InterviewService, InterviewException
from backend.middleware.auth_middleware import get_current_user

router = APIRouter(prefix="/interview", tags=["interview"])
interview_service = InterviewService()


@router.post("/start", response_model=InterviewStartResponse)
async def start_interview(
    body: InterviewStartRequest,
    current_user: UserProfile = Depends(get_current_user),
):
    result = interview_service.start_interview(
        user_id=current_user.id,
        resume_id=body.resume_id,
        target_role=body.target_role,
        interview_type=body.interview_type,
        max_questions=body.max_questions,  # ← add karo
    )
    return result


@router.post("/message", response_model=InterviewMessageResponse)
async def send_message(
    body: InterviewMessageRequest,
    current_user: UserProfile = Depends(get_current_user),
):
    result = interview_service.send_message(
        user_id=current_user.id,
        session_id=body.session_id,
        answer=body.answer,
    )
    return result


@router.get("/sessions", response_model=list[InterviewSessionSummary])
async def list_sessions(current_user: UserProfile = Depends(get_current_user)):
    return interview_service.get_sessions(current_user.id)


@router.get("/sessions/{session_id}", response_model=InterviewSessionDetail)
async def get_session(
    session_id: str,
    current_user: UserProfile = Depends(get_current_user),
):
    session = interview_service.get_session(current_user.id, session_id)
    session["messages"] = [InterviewMessage(**m) for m in session["messages"]]
    return session

File: frontend/lib/interview.ts
import api from "./axios"

export interface InterviewStartRequest {
  target_role: string
  interview_type: string
  max_questions?: number
  resume_id?: string
}

export interface InterviewStartResponse {
  session_id: string
  question: string
  question_type: string
  question_number: number
}

export interface InterviewMessageRequest {
  session_id: string
  answer: string
}

export interface InterviewFeedback {
  overall_score: number
  strengths: string[]
  weaknesses: string[]
  topic_wise_feedback: Record<string, string>
  sample_better_answers: Record<string, string>
  hiring_recommendation: "Strong Yes" | "Yes" | "No"
}

export interface InterviewMessageResponse {
  question: string | null
  question_type: string | null
  question_number: number | null
  interview_complete: boolean
  feedback: InterviewFeedback | null
}

export interface InterviewSessionSummary {
  id: string
  target_role: string
  interview_type: string
  status: string
  question_count: number
  max_questions: number
  created_at?: string
  completed_at?: string
}

export interface InterviewMessage {
  role: string
  content: string
  type: string
  question_number: number
}

export interface InterviewSessionDetail {
  id: string
  target_role: string
  interview_type: string
  max_questions: number
  status: string
  question_count: number
  messages: InterviewMessage[]
  feedback: InterviewFeedback | null
  created_at?: string
  completed_at?: string
}

export async function startInterview(
  data: InterviewStartRequest
): Promise<InterviewStartResponse> {
  const res = await api.post<InterviewStartResponse>("/interview/start", data)
  return res.data
}

export async function sendMessage(
  session_id: string,
  answer: string
): Promise<InterviewMessageResponse> {
  const res = await api.post<InterviewMessageResponse>("/interview/message", {
    session_id,
    answer,
  })
  return res.data
}

export async function listSessions(): Promise<InterviewSessionSummary[]> {
  const res = await api.get<InterviewSessionSummary[]>("/interview/sessions")
  return res.data
}

export async function getSession(
  session_id: string
): Promise<InterviewSessionDetail> {
  const res = await api.get<InterviewSessionDetail>(
    `/interview/sessions/${session_id}`
  )
  return res.data
}
